beets/test/fixtures.py

```diff
@@ -0,0 +1,107 @@
+# This file is part of beets.
+# Copyright 2024, Arav K.
+#
+# Permission is hereby granted, free of charge, to any person obtaining
+# a copy of this software and associated documentation files (the
+# "Software"), to deal in the Software without restriction, including
+# without limitation the rights to use, copy, modify, merge, publish,
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""
+This module provides `pytest`-based fixtures for testing Beets.
+"""
+
+from pathlib import Path
+from typing import Iterator
+
+import pytest
+from confuse import Configuration
+
+import beets
+from beets.library import Library
+
+
+@pytest.fixture(scope="session")
+def resource_dir(pytestconfig: pytest.Config) -> Path:
+    """
+    The resource directory for tests.
+
+    Tests requiring external data (e.g. audio files) can place them within the
+    resource directory (located at `test/rsrc` in the repository) and then find
+    them relative to the returned path.
+
+    If the tests for a particular component or plugin require several files,
+    they should be placed within an appropriately named subdirectory.
+
+    :return: the path to `test/rsrc` in the repository.
+    """
+
+    return pytestconfig.rootpath / "test" / "rsrc"
+
+
+@pytest.fixture
+def config(
```

Comment:
I'm a bit concerned about multiple sources of truth for the same beets config setup in `setup_beets` and this fixture. If these two configurations get out of sync, we are more likely to see unexpected test failures attempting to port tests from `unittest` to `pytest`. This can be a nightmare to debug due to the lack of clear indication of the root cause.

What would you think about defining the logic in your `config` fixture as a context manager in `beets.test.helper`? This way we could reuse it in both `BeetsTestCase` and `pytest` fixtures, and remove the duplicate implementation from `TestHelper.setup_beets`.

For example

```python
# beets/test/helper.py

@contextmanager
def setup_config(directory: Path):
    config = beets.config
    config.clear()
    config.read()

    config["plugins"] = []
    config["verbose"] = 1
    config["ui"]["color"] = False
    config["threaded"] = False
    config["directory"] = str(tmp_path)

    yield config

    self.config.clear()
    beets.config.read(user=False, defaults=True)
```

It encapsulates all setup and cleanup logic and it can be used both in the `setUp` code and `pytest` fixtures:

```python 
# beets/test/helper.py
class TestHelper:
    ...
    def setup_with_context_manager(self, cm):
        val = cm.__enter__()
        self.addCleanup(cm.__exit__, None, None, None)
        return val

    def setUp(self):
        ...
        self.config = self.setup_with_context_manager(setup_config(self.tmp_path))
        ...


# conftest.py
from beets.test.helper import setup_config

@pytest.fixture
def config(tmp_path):
    with setup_config(tmp_path) as config:
        yield config
```

This way we keep a single source of truth for the beets config setup, maintaining consistency regardless of the test runner. This pattern is also applicable to the rest of the setup (lib, importer etc.). What do you think?

A nice side effect of this approach is that we can use `conftest.py` to define our fixtures since the stuff useful to people is defined in `beets.test.helper` and not in the test case itself. This makes the test cases more readable and easier to maintain.

Reply:
I like this plan, thanks!  Once we're done with `unittest` we can revert back to the existing layout, though.  I'll leave a TODO in the code for it.